deadd-notification-center: 1.7.3 -> 2021-03-10 and service files for dbus and systemd

Pacman99 opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change
The service file for deadd-notification-center was mixed up. I use the systemd service file for dbus by mistake.

This adds both the systemd and dbus service files to the appropriate place and uses the right files from the source.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

jtojnar commented on pkgs/applications/misc/deadd-notification-center/default.nix:
```diff
@@ -34,13 +34,16 @@ stdenv.mkDerivation rec {
   ];
 
   installPhase = ''
-    mkdir -p $out/bin $out/share/dbus-1/services
+    mkdir -p $out/bin $out/lib/systemd $out/share/dbus-1/services
 
     cp $src/.out/${pname} $out/bin/
     chmod +x $out/bin/${pname}
 
-    sed "s|##PREFIX##|$out|g" $src/${pname}.service.in > \
+    sed "s|##PREFIX##|$out|g" $src/com.ph-uhl.deadd.notification.service.in > \
       $out/share/dbus-1/services/com.ph-uhl.deadd.notification.service
+
+    sed "s|##PREFIX##|$out|g" $src/${pname}.service.in > \
+      $out/lib/systemd/${pname}.service
   '';
```
Why not use the `make install` in the package, using something like:

```patch
diff --git a/pkgs/applications/misc/deadd-notification-center/default.nix b/pkgs/applications/misc/deadd-notification-center/default.nix
index 35627b28d15..839a3228a0b 100644
--- a/pkgs/applications/misc/deadd-notification-center/default.nix
+++ b/pkgs/applications/misc/deadd-notification-center/default.nix
@@ -19,8 +19,6 @@ stdenv.mkDerivation rec {
     sha256 = "QaOLrtlhQyhMOirk6JO1yMGRrgycHmF9FAdKNbN2TRk=";
   };
 
-  dontUnpack = true;
-
   nativeBuildInputs = [
     autoPatchelfHook
     wrapGAppsHook
@@ -33,15 +31,18 @@ stdenv.mkDerivation rec {
     hicolor-icon-theme
   ];
 
-  installPhase = ''
-    mkdir -p $out/bin $out/share/dbus-1/services
-
-    cp $src/.out/${pname} $out/bin/
-    chmod +x $out/bin/${pname}
+  buildFlags = [
+    # Exclude stack from `make all` to use the prebuilt binary from `.out/`.
+    "service"
+  ];
 
-    sed "s|##PREFIX##|$out|g" $src/${pname}.service.in > \
-      $out/share/dbus-1/services/com.ph-uhl.deadd.notification.service
-  '';
+  makeFlags = [
+    "PREFIX=${placeholder "out"}"
+    "SERVICEDIR_SYSTEMD=${placeholder "out"}/etc/systemd/user"
+    "SERVICEDIR_DBUS=${placeholder "out"}/share/dbus-1/services"
+    # Override systemd auto-detection.
+    "SYSTEMD=1"
+  ];
 
   meta = with lib; {
     description = "A haskell-written notification center for users that like a desktop with style";
```

Though ideally, we would build it from scratch.

Pacman99 replied:
Yeah that is much cleaner, thanks for the suggestion, I applied the patch.

I would still like to build from scratch, the main problem is broken dependencies in haskellPackages.

teto commented:
I've got
```
   services.dbus.packages = [
      pkgs.deadd-notification-center
    ];

  systemd.packages = [
    pkgs.deadd-notification-center
  ];
```
yet `systemctl list-units` doesn't list deadd ?
As for the dbus part as I am using home-manager so I am not sure if deadd  is supposed to work in the first place, I probably would need to tweak things a bit.

Pacman99 commented:
I think the service gets placed under etc/systemd/user so it would show under systemctl --user. I'm not sure how user services work with home-manager. You might have to write the service file manually for that.

teto commented:
Indeed didn't see the `.../user/...` . I tried  `systemctl --user start deadd-notification-center.service` which  fails here
```
Feb 17 19:53:40 mcoudron systemd[1691]: Stopped Deadd Notification Center.
Feb 17 19:53:40 mcoudron systemd[1691]: Started Deadd Notification Center.
Feb 17 19:53:40 mcoudron deadd-notification-center[28885]: notificationDaemon started
Feb 17 19:53:40 mcoudron deadd-notification-center[28885]: deadd-notification-center: locale: createProcess: runInteractiveProcess: exec: does not exist (No such file or directory)
Feb 17 19:53:40 mcoudron systemd[1691]: deadd-notification-center.service: Main process exited, code=exited, status=1/FAILURE
Feb 17 19:53:40 mcoudron systemd[1691]: deadd-notification-center.service: Failed with result 'exit-code'.
```
wonder if the issue is linked with `locale`. Does it work on your side ?

Pacman99 commented:
I don't actually run deadd-notification-center via systemd, so I usually just test the package by running the binary. 
But when I tried with systemctl I get a different error:
```
Unable to init server: Could not connect: Connection refused
cannot open display: :0
```
I'm not super proficient with systemd service files, so I don't yet know the fix to that.

Commit:
deadd-notification-center: 1.7.3 -> 2021-03-10

Commit:
deadd-notification-center: use make to install and fix issues with service files

Pacman99 commented:
I found this patch phuhl/linux_notification_center#121 that should fix the system-locale issue. @teto Your problem was reported upstream in phuhl/linux_notification_center#63 and that patch should fix it. But it didn't apply cleanly to 1.7.3, so I just updated the package to the latest commit. It would be great if you could test it, since I wan't able to reproduce your issue.

And to fix my issue I added a patch from an open PR to simplify the dbus and systemd service files.

teto commented:
thanks. I see an error in the patch hash. Even after fixing it it stills fails for me maybe 
```
error: hash mismatch in fixed-output derivation '/nix/store/kfif8wpcay72q3byqqm0x8v8y2rm7rq9-5244e1498574983322be97925e1ff7ebe456d974.patch.drv':
         specified: sha256-5LuJ/C7VRcg1ndBmplCCxrVo3YVXl30HKnU+4dfPbMU=
            got:    sha256-hbqbgBmuewOhtx0na2tmFa5W128ZrBvDcyPme/mRzlI=
```
which is weird since the env-locale change should fix that :s I will try after a reboot.

Pacman99 commented:
Huh thats wierd thats the hash `nix-prefetch-url` gave for me. But I updated the hash to what you sent.

jtojnar commented:
`fetchpatch` cleans the patch a bit and only the result is hashed. `nix-prefetch-url` will return the hash of the unclean patch.

teto commented on pkgs/applications/misc/deadd-notification-center/default.nix:
```diff
@@ -7,19 +7,25 @@
 stdenv.mkDerivation rec {
   pname = "deadd-notification-center";
-  version = "1.7.3";
+  version = "2020-03-10";
```
2021 probably

Pacman99 replied:
ahh right of course, fixed

teto commented:
doesn't work for me but looks like an improvement already. Will get easier to fix from there thus merging. Thanks for the help.

Pacman99 commented:
I wonder if the env-locale fixes never got built. Because it looks like the .out folder in the repo was only updated 2 months ago. I'm not entirely sure out the github actions workflow works.

teto commented:
I suspect sthg similar. Didn't know about the `.out` but that must be it.

teto commented:
also worth noting I had to create a deadd.css file also (https://github.com/phuhl/linux_notification_center/issues/132)
